opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1808,56 +1808,62 @@ namespace Opm {
-            updateAndCommunicate(episodeIdx, iterationIdx, deferred_logger);
-            changed_well_group = true;
-        }
+            changed_well_to_group = comm.sum(static_cast<int>(changed_well_to_group));
+            if (changed_well_to_group) {
+                updateAndCommunicate(episodeIdx, iterationIdx, deferred_logger);
+                changed_well_group = true;
+            }
 
-        // Check individual well constraints and communicate.
-        bool changed_well_individual = false;
-        {
-            // For MS Wells a linear solve is performed below and the matrix might be singular.
-            // We need to communicate the exception thrown to the others and rethrow.
-            OPM_BEGIN_PARALLEL_TRY_CATCH()
-                for (const auto& well : well_container_) {
-                    const auto mode = WellInterface<TypeTag>::IndividualOrGroup::Individual;
-                    const bool changed_well = well->updateWellControl(simulator_, mode, this->wellState(), this->groupState(), deferred_logger);
-                    if (changed_well) {
-                        changed_well_individual = changed_well || changed_well_individual;
+            // Check individual well constraints and communicate.
+            bool changed_well_individual = false;
+            {
+                // For MS Wells a linear solve is performed below and the matrix might be singular.
+                // We need to communicate the exception thrown to the others and rethrow.
+                OPM_BEGIN_PARALLEL_TRY_CATCH()
+                    for (const auto& well : well_container_) {
+                        const auto mode = WellInterface<TypeTag>::IndividualOrGroup::Individual;
+                        const bool changed_well = well->updateWellControl(simulator_, mode, this->wellState(), this->groupState(), deferred_logger);
+                        if (changed_well) {
+                            changed_well_individual = changed_well || changed_well_individual;
+                        }
                     }
-                }
-            OPM_END_PARALLEL_TRY_CATCH("BlackoilWellModel: updating well controls failed: ",
-                                       simulator_.gridView().comm());
-        }
+                OPM_END_PARALLEL_TRY_CATCH("BlackoilWellModel: updating well controls failed: ",
+                                        simulator_.gridView().comm());
+            }
 
-        changed_well_individual = comm.sum(static_cast<int>(changed_well_individual));
-        if (changed_well_individual) {
-            updateAndCommunicate(episodeIdx, iterationIdx, deferred_logger);
-            changed_well_group = true;
+            changed_well_individual = comm.sum(static_cast<int>(changed_well_individual));
+            if (changed_well_individual) {
+                updateAndCommunicate(episodeIdx, iterationIdx, deferred_logger);
+                changed_well_group = true;
+            }
+            iter++;
```

Comment:
What should the practice for this type of loop, do we need to use comm.barrier() here at the end of the loop?

@blattms @akva2